File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.java
class Solution {
    public String shortestBeautifulSubstring(String s, int k) {
        int i=0;
        int count = 0;
        
        int best = Integer.MAX_VALUE;
        String ans = "";
        
        for (int j=0; j<s.length(); j++) {
            if (s.charAt(j) == '1') count++;

            while (count == k) {
                int len = j - i + 1;

                // Java uses [start, end)
                // unlike C++, which uses [start, start + length).
                String current = s.substring(i, j+1);

                if (len < best || (len == best && current.compareTo(ans) < 0)) {
                    best = len;
                    ans = current;
                }

                if (s.charAt(i) == '1') count--;

                i++;
            }
        }

        return ans;
    }
}

File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.cpp
class Solution {
public:
    string shortestBeautifulSubstring(string s, int k) {
        int i=0;
        int count = 0;
        
        int best = INT_MAX;
        string ans = "";
        
        for (int j=0; j<s.size(); j++) {
            if (s[j] == '1') count++;

            while (count == k) {
                int len = j - i + 1;

                string current = s.substr(i, len);

                if (len < best || (len == best && current < ans)) {
                    best = len;
                    ans = current;
                }

                if (s[i] == '1') count--;

                i++;
            }
        }

        return ans;
    }
};